### PR TITLE
Refactored contract wrapper

### PR DESCRIPTION
**Pull request by DariuszDepta**
Refactored the `ContractWrapper`, added some documentation, to shed some light on the purpose of used generic types.

### CODE REVIEW

File: src/contracts.rs [hunk 1]
```diff
@@ -1,70 +1,126 @@
...
+    pub type QueryClosure<T, E, Q> = Box<dyn Fn(Deps<Q>, Env, T) -> Result<Binary, E>>;
+}
+
+use closures::*;
+
+/// This structure wraps the [Contract] trait implementor
+/// and provides generic access to the contract's entry-points.
+///
+/// List of generic types used in [ContractWrapper]:
+/// - **T1** type of message passed to [execute] entry-point.
```
Comment (kulikthebird) [inline src/contracts.rs:61]: I like the new documentation added to this part of multitest code. I'd propose one more thing on top of that. What if we change all those one letter generics into proper names that acually inform users about their purpose? For e.g. `T1` -> `ExecuteMsgType`, `E1` -> `ExecuteError`

File: src/contracts.rs [hunk 2]
```diff
@@ -382,20 +444,23 @@ where
...
-    T6: DeserializeOwned,
-    E1: Display + Debug + Send + Sync + Error + 'static,
-    E2: Display + Debug + Send + Sync + Error + 'static,
-    E3: Display + Debug + Send + Sync + Error + 'static,
-    E4: Display + Debug + Send + Sync + 'static,
-    E5: Display + Debug + Send + Sync + 'static,
-    E6: Display + Debug + Send + Sync + 'static,
-    C: CustomMsg,
-    Q: CustomQuery + DeserializeOwned,
+    T1: DeserializeOwned + 'static, // Type of message passed to `execute` entry-point.
```
Comment (kulikthebird) [inline src/contracts.rs:447]: The less we are strict in the generics bounds the better. This additional `'static` in a trait's types is not necessary.

File: src/contracts.rs [hunk 3]
```diff
@@ -1,70 +1,126 @@
...
+/// Response<**C**> or **Binary** and an error of type **E1**, **E2**, **E3**, **E4**, **E5**
+/// or **E6**.
+///
+/// ```text
+///    entry_point(query, .., message_in) -> Result<message_out, error>
+///                  ┬           ┬                      ┬          ┬
+///             Q >──┘           │                      │          └──> E1,E2,E3,E4,E5,E6
+///    T1,T2,T3,T4,Reply,T6 >────┘                      └─────────────> C,Binary
+/// ```
+/// Generic type **C** defines a custom message that specific for the **whole blockchain**.
```
Comment (kulikthebird) [inline src/contracts.rs:111]:
```suggestion
/// Generic type **C** defines a custom message that is specific for the **whole blockchain**.
```

File: src/contracts.rs [hunk 4]
```diff
@@ -1,70 +1,126 @@
...
+/// Similarly, the generic type **Q** defines a custom query that is also specific
+/// to the **whole blockchain**. Other generic types a specific to the implemented contract.
```
Comment (kulikthebird) [inline src/contracts.rs:113]:
```suggestion
/// to the **whole blockchain**. Other generic types are specific to the implemented contract.
```

File: src/contracts.rs [hunk 5]
```diff
@@ -1,70 +1,126 @@
...
+/// So all smart contracts used in the same blockchain will have the same types for **C** and **Q**,
+/// but each contract may use different type for other generic types.
+/// It means that e.g. **T1** in smart contract `A` may differ from **T1** in smart contract `B`.
```
Comment (kulikthebird) [inline src/contracts.rs:116]: Let's add a link to the cosmwasm documentation that describes the custom message types here

**Review by kulikthebird — CHANGES_REQUESTED** [review]
I like this change. Let's discuss some proposal in the comments before the merge.

File: src/contracts.rs (hunk 2 above)
Comment (DariuszDepta) [reply src/contracts.rs]: Removed `'static` from trait type's bounds.

File: src/contracts.rs (hunk 3 above)
Comment (DariuszDepta) [reply src/contracts.rs]: Fixed.

File: src/contracts.rs (hunk 4 above)
Comment (DariuszDepta) [reply src/contracts.rs]: Fixed.

File: src/contracts.rs (hunk 5 above)
Comment (DariuszDepta) [reply src/contracts.rs:116]: Custom messages and queries are specific to every blockchain, so it is hard to find a single source of truth.

File: src/contracts.rs (hunk 1 above)
Comment (DariuszDepta) [reply src/contracts.rs:61]: I would stick to short names for now, just for brevity and wait for the **"big"** refactor of this moidule.

**Review by kulikthebird — APPROVED** [review]

**Pull request merged**